IndexUpdatePlannerがBlockをオーバーフローする

yokomotod opened the pull request:
```
--- FAIL: TestGroupByPlan (0.01s)
panic: runtime error: slice bounds out of range [:404] with capacity 400 [recovered]
	panic: runtime error: slice bounds out of range [:404] with capacity 400

goroutine 35 [running]:
testing.tRunner.func1.2({0x702a80, 0xc0001672f0})
	/opt/hostedtoolcache/go/1.22.3/x64/src/testing/testing.go:1631 +0x3f7
testing.tRunner.func1()
	/opt/hostedtoolcache/go/1.22.3/x64/src/testing/testing.go:1634 +0x6b6
panic({0x702a80?, 0xc0001672f0?})
	/opt/hostedtoolcache/go/1.22.3/x64/src/runtime/panic.go:770 +0x132
simpledb/file.(*Page).GetInt(...)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/file/file.go:48
simpledb/tx/recovery.(*Manager).SetInt(0xc000196210, 0xc000120cc0, 0x190, 0xc0?)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/tx/recovery/recovery.go:77 +0x2e5
simpledb/tx.(*Transaction).SetInt(0xc00019[61](https://github.com/yokomotod/database-design-and-implementation-go/actions/runs/10898524763/job/30241922151?pr=30#step:4:62)e0, {{0xc0001b3b71, 0xf}, 0x0}, 0x190, 0x1e, 0x1)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/tx/transaction.go:114 +0x15b
simpledb/index/btree.(*BTreePage).setInt(0xc0001b68a0, 0x6, {0x7190d5, 0x7}, 0x1e)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/index/btree/page.go:124 +0x17a
simpledb/index/btree.(*BTreePage).setVal(0xc0001b68a0, 0x6, {0x7190d5, 0x7}, 0xc000109530)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/index/btree/page.go:139 +0x34d
simpledb/index/btree.(*BTreePage).copyRecord(0xc0001b68a0, 0x5, 0x6)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/index/btree/page.go:177 +0x126
simpledb/index/btree.(*BTreePage).insert(0xc0001b68a0, 0x0)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/index/btree/page.go:161 +0xe5
simpledb/index/btree.(*BTreePage).InsertLeaf(0xc0001b68a0, 0x0, 0xc000109410, 0xc0001b3688)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/index/btree/page.go:317 +0x49
simpledb/index/btree.(*BTreeLeaf).Insert(0xc000121e80, 0xc0001b3688)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/index/btree/leaf.go:136 +0x78d
simpledb/index/btree.(*BTreeIndex).Insert(0xc00014a550, 0xc000109410, 0xc0001b3688)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/index/btree/index.go:123 +0x94
simpledb/plan.(*IndexUpdatePlanner).ExecuteInsert(0xc00010a0d0, 0xc000121d40, 0xc0001961e0)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/plan/index_update_planner.go:58 +0x6f6
simpledb/plan.(*Planner).ExecuteUpdate(0xc0001444c0, {0xc0001[64](https://github.com/yokomotod/database-design-and-implementation-go/actions/runs/10898524763/job/30241922151?pr=30#step:4:65)640, 0x4d}, 0xc0001961e0)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/plan/planner.go:70 +0x222
simpledb/testlib.InsertTestData(0xc00014e9c0, 0xc00010cff0)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/testlib/helper.go:70 +0x69c
simpledb/plan_test.TestGroupByPlan(0xc00014e9c0)
	/home/runner/work/database-design-and-implementation-go/database-design-and-implementation-go/simpledb/plan/group_by_plan_test.go:17 +0x191
testing.tRunner(0xc00014e9c0, 0x742cc0)
	/opt/hostedtoolcache/go/1.22.3/x64/src/testing/testing.go:1[68](https://github.com/yokomotod/database-design-and-implementation-go/actions/runs/10898524763/job/30241922151?pr=30#step:4:69)9 +0x21f
created by testing.(*T).Run in goroutine 1
	/opt/hostedtoolcache/go/1.22.3/x64/src/testing/testing.go:1742 +0x826
FAIL	simpledb/plan	0.026s
```

yokomotod committed:
fix IndexUpdatePlanner interface

yokomotod committed:
try to use IndexUpdatePlanner

yokomotod commented:
https://github.com/yokomotod/database-design-and-implementation-go/blob/3f8dddfde0f506400f75ca2cd042dd09f317e3e0/simpledb/index/btree/leaf.go#L106-L111

このあたりの処理・・・か・・・・？

adieumonks commented:
bree indexのtretransferRecsがバグってます

adieumonks committed:
fix btree page

adieumonks commented:
@yokomotod https://github.com/yokomotod/database-design-and-implementation-go/pull/31 で直ると思います

yokomotod committed:
Merge branch 'fix/btree-page' into fix-index-update-planner

yokomotod commented:
```
--- FAIL: TestStudent (0.00s)
    planner_test.go:26: created table student
    planner_test.go:32: insert into student: 1
    planner_test.go:41: insert into student: 1
    planner_test.go:65: deleted count: want: 0, got: 3
    planner_test.go:73: deleted count: want: 1, got: 0
```

https://github.com/yokomotod/database-design-and-implementation-go/blob/6aec99c59298771a76cd0aa41834bd41a1ff686c/simpledb/plan/planner_test.go#L60-L74

@adieumonks まだコケる部分が残ってますが、改善してますね！！

yokomotod committed:
fix IndexUpdatePlanner.ExecuteDelete

yokomotod commented:
https://github.com/yokomotod/database-design-and-implementation-go/pull/30/commits/c30acd23ff08535908c4b406a930143f00f67a1e

治った

修正用PR → #32

yokomotod commented:
#31 #32 で修正done
